# origin: marcelo (BitKeeper)
# cset: 1.1448.1.134 (2.4) key=4238824fdffedsyIDAkzZNeidxjirw
# URL: http://linux.bkbits.net:8080/linux-2.4/cset@4238824fdffedsyIDAkzZNeidxjirw
# inclusion: upstream
# descrition: Andrea Arcangeli: get_user_pages() shall not grab PG_reserved pages
# revision date: Thu, 14 Apr 2005 11:59:24 +0900
#
# S rset: ChangeSet|1.1448.1.133..1.1448.1.134
# I rset: mm/memory.c|1.64..1.65
#
# Key:
# S: Skipped  ChangeSet file only
# O: Original Followed by Updated
# U: Updated  Included with updated range of versions
# I: Included Included verbatim
# E: Excluded Excluded on request from user
# D: Deleted  Manually deleted by subsequent user edit
# R: Revised  Manually revised by subsequent user edit
#
#
# This is a BitKeeper generated diff -Nru style patch.
#
# ChangeSet
#   2005/03/16 12:25:52-03:00 marcelo@logos.cnet 
#   Andrea Arcangeli: get_user_pages() shall not grab PG_reserved pages
# 
# mm/memory.c
#   2005/03/16 12:23:35-03:00 marcelo@logos.cnet +5 -3
#   Andrea Arcangeli: get_user_pages() shall not grab PG_reserved pages
# 
#
===== mm/memory.c 1.64 vs 1.65 =====
--- 1.64/mm/memory.c	2005-01-11 21:43:40 +09:00
+++ 1.65/mm/memory.c	2005-03-17 00:23:35 +09:00
@@ -499,9 +499,11 @@ int get_user_pages(struct task_struct *t
 				/* FIXME: call the correct function,
 				 * depending on the type of the found page
 				 */
-				if (!pages[i])
-					goto bad_page;
-				page_cache_get(pages[i]);
+				if (!pages[i] || PageReserved(pages[i])) {
+					if (pages[i] != ZERO_PAGE(start))
+						goto bad_page;
+				} else
+					page_cache_get(pages[i]);
 			}
 			if (vmas)
 				vmas[i] = vma;
